Stricter C99 warnings, generate primitive list with types

MisterDA opened the pull request:
I was working on this at the end of last week after reading [Modernizing Fedora's C code](https://lwn.net/Articles/913505/), and over the week-end #11759 was opened which this PR should fix.

This PR:
- enables new warning for stricter C99 generation;
- fixes strict-prototypes and old-style-definition warnings;
- generates the list of primitives with their argument list.

It generates the list of primitives with their arguments by running first the C preprocessor on C files containing primitives, extracting the functions with a sed script. It retains the caching behaviour introduced in #8696. It also introduces `PrimitiveX` macros for all primitive arities.

Possible follow-ups for this PR include checking support from the compiler for the warnings, defining a better `CAMLprim` macro _declaring_ the arity of the function and possibly using it at runtime to check the arity of a called function.

MisterDA committed:
Enable stricter C99 warnings

Fedora is considering moving all its packages to C99 and enabling
stricter C99 warnings. OCaml has already moved to a C11 compiler,
let's enable these warnings on GCC-compatible compilers.

Some of the warnings on the following article are already covered by
-Wall and -Werror.

Reference: [Modernizing Fedora's C code][1].

[1]: https://lwn.net/Articles/913505/

fixup

MisterDA committed:
Generate the list of primitives using the preprocessor

The new code uses the preprocessor to expand possible macros in the
generation of the primitive prototype (for instance, some comparison
functions in runtime/floats.c). The preprocessor will however remove
the token CAMLprim, so we redefine in the call CAMLprim to
CAMLprim()=0, as a function-like macro will not be expanded if not
given any arguments.

The list of primitives is kept sorted to benefit from stability if it
doesn't change and functions just moved around.

A sed script was extracted in a separate file to simplify quoting and
escaping in the Makefile.

The caching mechanism of runtime/primitives is preserved: if the list
of primitives hasn't changed, this file is not modified.

MisterDA committed:
Generate the list of primitives with accurate types

sed scripts are extracted in files in order to simplify quoting and
escaping in the Makefile.

It is always possible to cast a function pointer to another function
pointer type, and cast again to the original type before calling the
pointed function. I pick the base

    typedef value (*c_primitive)(void);

to store the functions in the array, and cast them back in the
interpretor according to the arity described by the opcode.

Some reference:
- [Function pointers in C][1]
- [Function pointer casts][2]

[1]: https://frama-c.com/2013/08/24/Function-pointers-in-C.html
[2]: https://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf

MisterDA commented:
@xavierleroy was faster than me with #11762 which is equivalent to my 54210b5b.

xavierleroy commented:
Could you please submit 54210b5bf3 (the middle commit) as a separate PR, so that it can be merged immediately?  (Plus, it subsumes my PR #11762, so I'll be able to close #11762.)

xavierleroy commented:
I think your 54210b5 is more comprehensive than my #11762, so let's use your code instead.

MisterDA commented:
Submitted this commit as #11764. I've used your commit message instead of mine.

MisterDA committed:
Fix an old-style-declaration warning

> Warn if storage-class specifiers like static are not the first
> things in a declaration.

dra27 commented:
At the moment, this goes to considerable lengths to generate accurate prototypes for the primitives. Those prototypes are then forcibly cast to `c_primitive` in order to put them all in `caml_builtin_cprim`. Wouldn't it be easier just to declare them all as `caml_prim(void)` in `prims.c`? There's a tiny gotcha because `runtime/caml/mlvalues.h` contains the actual prototypes for `caml_set_oo_name` and `caml_get_public_method` but this can be relatively easier worked around to get this patch which when combined with just `runtime/interp.c` and `runtime/caml/prims.h` from 63cc8272c7e2723b801c98749accaed1cdf7a279 also builds without triggering any warnings:

```diff
--- a/Makefile
+++ b/Makefile
@@ -771,10 +771,10 @@ runtime/primitives: \
        cp $^ $@

 runtime/prims.c : runtime/primitives
-       (echo '#define CAML_INTERNALS'; \
-         echo '#include "caml/mlvalues.h"'; \
-        echo '#include "caml/prims.h"'; \
-        sed -e 's/.*/extern value &();/' $<; \
+       (echo '#include "caml/config.h"'; \
+        echo 'typedef intnat value;'; \
+        echo 'typedef value (*c_primitive)(void);'; \
+        sed -e 's/.*/extern value &(void);/' $<; \
         echo 'c_primitive caml_builtin_cprim[] = {'; \
         sed -e 's/.*/  &,/' $<; \
         echo '  0 };'; \
```

xavierleroy commented on Makefile:
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
+  $(runtime_PRIMITIVES) \
+  runtime/gen_primitives.sed runtime/gen_primitives_ints.sed
+	(for prim in $(runtime_PRIMITIVES); do \
+	     $(CPP) -I./runtime -D'CAMLprim()=0' "$${prim}" | \
+	     sed -n -f runtime/gen_primitives.sed; \
```
Is the preprocessing really needed?

MisterDA replied:
The preprocessor is needed for some symbols in `runtime/floats.c`:

https://github.com/ocaml/ocaml/blob/d0cdf4b0a075ddb56a5523c4d51c99408321f609/runtime/floats.c#L1127-L1131

I couldn't extract the arguments without expanding the macros. Using the preprocessor also removes this concern (I forgot to remove the comment):

```
# If primitives contain duplicated lines (e.g. because the code is defined
# like
# #ifdef X
# CAMLprim value caml_foo() ...
# #else
# CAMLprim value caml_foo() ...
# #endif)
```

xavierleroy replied:
Good points, thanks for the explanations.  The `DEFINE_NAN_CMP` macro in `runtime/floats.c` is ugly as hell and could be rewritten.  For the alternate definitions, we used to rely on `sort | uniq` to keep only one, but it gets more fragile when the prototypes are taken into account, that much is true.

xavierleroy commented on runtime/caml/prims.h:
```diff
@@ -30,7 +30,15 @@ extern struct ext_table caml_prim_table;
 #endif
 
-#define Primitive(n) ((c_primitive)(caml_prim_table.contents[n]))
+#define Primitive1(n) ((value (*)(value))caml_prim_table.contents[n])
+#define Primitive2(n) ((value (*)(value,value))caml_prim_table.contents[n])
+#define Primitive3(n) \
+    ((value (*)(value,value,value))caml_prim_table.contents[n])
+#define Primitive4(n) \
+    ((value (*)(value,value,value,value))caml_prim_table.contents[n])
+#define Primitive5(n) \
+    ((value (*)(value,value,value,value,value))caml_prim_table.contents[n])
+#define PrimitiveN(n) ((value (*)(value*,int))caml_prim_table.contents[n])
```
Do yourself a favor and introduce typedefs:
```
typedef value (*c_primitive1)(value);
...
#define Primitive1(n) (c_primitive1)(caml_prim_table.contents[n]))
```

MisterDA replied:
That's much more readable, thanks.

xavierleroy commented on runtime/caml/prims.h:
```diff
@@ -20,7 +20,7 @@
 #ifdef CAML_INTERNALS
 
-typedef value (*c_primitive)();
+typedef value (*c_primitive)(void);
```
You could just as well use
```
typedef void * c_primitive;
```
The built-in primitives from table `caml_builtin_cprim` are never called directly anyway, just added to an extensible table with type `void *`.

(Note that ISO C forbids casting a function pointer to type `void *`, but we've been doing this for a very long time, and all C compilers that I know of are happy with this.)

MisterDA replied:
Also nice, thanks.

xavierleroy commented on runtime/gen_primitives.sed:
```diff
@@ -15,22 +13,13 @@
 #**************************************************************************
 
-# duplicated from $(ROOTDIR)/runtime/Makefile
+/^CAMLprim value/ {
+  :loop
+  /)/ ! {
+    N; b loop
+  }
+
+  s/\n  */ /g
```
I didn't know you had conditional jumps in sed...  This seems to be in POSIX sed, even.  

This said, if you're willing to suppose primitive definitions take at most 2 lines, there's a loop-free approach:
on seeing `/^CAMLprim /`, fetch next line with `N`, then remove everything following the first `)`.

MisterDA replied:
I could, but `caml_array_fill` would disagree!
https://github.com/ocaml/ocaml/blob/ac52c3ab4a545d77eea034993a0947ba5e79762b/runtime/array.c#L557-L560
I think there might be fewer surprises (apart from the script itself) if we loop in sed.

xavierleroy commented on runtime/sys.c:
```diff
@@ -219,7 +219,7 @@ CAMLprim value caml_sys_exit(value retcode)
 #endif
 
-const static int sys_open_flags[] = {
+static const int sys_open_flags[] = {
```
Unrelated change.

MisterDA replied:
This fixes the `-Wold-style-declaration` warning that I couldn't enable globally, as it's specific to GCC.

xavierleroy commented on Makefile:
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
+	 echo; \
 	 echo 'c_primitive caml_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  &,/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  (c_primitive)\1,/' $<; \
```
With my suggestion `typedef void * c_primitive;`, the cast to `(c_primitive)` is not needed.

MisterDA replied:
Applied, thanks.

xavierleroy commented on Makefile:
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
+	 echo; \
 	 echo 'char * caml_names_of_builtin_cprim[] = {'; \
-	 sed -e 's/.*/  "&",/' $<; \
+	 sed -e 's/^\([a-z0-9_][a-z0-9_]*\) *(.*)/  "\1",/' $<; \
```
All this `sed` without setting `LC_CTYPE` makes me nervous.  Experience shows that it fails with bizarre locales.  Or, use named character classes, e.g. `[[:alnum]_]`.

MisterDA replied:
The comment above suggested `LC_ALL=C` instead of `LC_CTYPE`. I'm not familiar with these, which one should I use? In the meantime, I've using `export LC_ALL=C;` where needed.

xavierleroy commented on Makefile:
```diff
@@ -762,24 +769,43 @@ runtime/ld.conf: $(ROOTDIR)/Makefile.config
-                    cmp -s runtime/primitives runtime/primitives.new || \
-                    echo runtime/primitives.new)
-	cp $^ $@
+# To speed up builds, we avoid changing "primitives" and
+# "primitive-arguments" when files containing primitives change but
+# the primitives table does not
+runtime/primitives-arguments.new: \
+  $(runtime_BUILT_HEADERS) \
+  $(runtime_PRIMITIVES) \
```
You're moving code from a script into the toplevel Makefile.  I'm not sure this is a good idea.

MisterDA replied:
It should be ok: we switch from one trick to use a script to conditionally add a prerequisite or not, which felt like advanced magic, to resetting the modified time of a file.
It does break the Dune build, though.

xavierleroy replied:
The Dune build could be a concern.  But my more general concern is that too much scripting within a Makefile renders the Makefile unreadable; at some point, having separate scripts that do one thing but do it well becomes preferable.

MisterDA replied:
I agree, that's why I've externalised the sed scripts.
However if we chose to keep using the preprocessor, I think it'll be difficult to give the `$(CPP)` content to the script due to usual concerns of spaces and word-splitting.

xavierleroy reviewed (commented):
Some suggestions and remarks.  I admire your mastery of `sed`, but would still prefer a more civilized scripting language.

MisterDA committed:
Add typedefs for primitive types

MisterDA committed:
Remove outdated comment

Using the preprocessor to extract the list of primitives solves the
issue raised by the example.

MisterDA committed:
Set LC_ALL=C to protect sed calls

MisterDA committed:
Store primitives as simple void pointers

> Note that ISO C forbids casting a function pointer to type void *,
> but we've been doing this for a very long time, and all C compilers
> that I know of are happy with this.

(comment by Xavier Leroy)

This is documented by C11 Standard J.5.7, Function pointer casts.

xavierleroy commented:
Concerning @dra27's suggestion:

> At the moment, this goes to considerable lengths to generate accurate prototypes for the primitives. Those prototypes are then forcibly cast to c_primitive in order to put them all in caml_builtin_cprim. Wouldn't it be easier just to declare them all as caml_prim(void) in prims.c? There's a tiny gotcha because runtime/caml/mlvalues.h contains the actual prototypes for caml_set_oo_name and caml_get_public_method but this can be relatively easier worked around […]

On the one hand, it's not legal C11 to declare `value f(void)` a function `f` that is defined in another compilation unit as `value f(value)`, say; while I think it is correct to declare it `value f()`, like we currently do.  On the other hand, this is no worse than the many other "not legal C11" things we already do,  and it does work around the problem with great economy of means.  So maybe we should commit this solution (plus the modified prims.h and interp.c files) first.

MisterDA commented:
Thank you both for these rounds of comprehensive reviews. As it seems that keeping the primitives prototypes is not useful right now (which is on me, I may have dove to deep in sed and Makefile scripting), I'm closing this PR. I'll submit a simpler one soon, without the full prototypes and the preprocessor pass. We can always revisit this idea later.
